go/pkg/cas/upload.go

```diff
@@ -589,6 +604,144 @@ func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlob
+
+	// TODO(nodir): implement per-RPC timeouts. No nice way to do it.
+	rewind := false
+	return u.withRetries(ctx, func(ctx context.Context) error {
+		// Do not rewind if this is the first attempt.
+		if rewind {
+			if _, err := r.Seek(0, io.SeekStart); err != nil {
+				return err
+			}
+		}
+		rewind = true
+
+		// Start the byte stream.
+		stream, err := u.byteStream.Write(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Compress using an in-memory pipe. This is mostly to accomodate the fact
+		// that zstd package expects a writer.
+		// Note that using io.Pipe() means we buffer only bytes that were not uploaded yet.
+		pr, pw := io.Pipe()
+		enc.Reset(pw)
+
+		// Read from disk and make RPCs concurrently.
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
+			switch _, err := io.CopyBuffer(enc, r, fileBuf); {
+			case err == io.ErrClosedPipe:
+				// The other goroutine exited before we finished encoding.
+				// Might be a cache hit or context cancelation.
+				// In any case, the other goroutine has the actual error, so return nil
+				// here.
+				return nil
+			case err != nil:
+				return errors.Wrapf(err, "failed to read the file/blob")
+			}
+
+			if err := enc.Close(); err != nil {
+				return errors.Wrapf(err, "failed to close the zstd encoder")
+			}
+			return pw.Close()
+		})
+		eg.Go(func() error {
+			defer pr.Close()
+			req := &bspb.WriteRequest{
+				ResourceName: fmt.Sprintf("%s/uploads/%s/compressed-blobs/zstd/%s/%d", u.InstanceName, uuid.New(), item.Digest.Hash, item.Digest.SizeBytes),
+			}
+		chunkLoop:
+			for {
+				// Before reading, check if the context if canceled.
+				if ctx.Err() != nil {
+					return ctx.Err()
+				}
+
+				// Read the next chunk from the pipe.
+				// Use ReadFull to ensure we aren't sending tiny blobs over RPC.
+				n, err := io.ReadFull(pr, streamBuf)
+				switch {
+				case err == io.EOF || err == io.ErrUnexpectedEOF:
+					req.FinishWrite = true
+				case err != nil:
+					return err
+				}
+				req.Data = streamBuf[:n] // note: must happen in ErrUnexpectedEOF case
+
+				// Send the chunk.
+				switch err = stream.Send(req); {
+				case err == io.EOF:
+					// The server closed the stream.
+					// Most likely the file is already uploaded, see the CommittedSize check below.
+					break chunkLoop
+				case err != nil:
+					return err
+				case req.FinishWrite:
+					break chunkLoop
+				}
+
+				// Prepare the next request.
+				req.ResourceName = "" // send the resource name only in the first request
+				req.WriteOffset += int64(len(req.Data))
+			}
+
+			// Finalize the request.
+			switch res, err := stream.CloseAndRecv(); {
+			case err != nil:
+				return err
+			case res.CommittedSize != item.Digest.SizeBytes:
+				return fmt.Errorf("unexpected commitSize: got %d, want %d", res.CommittedSize, item.Digest.SizeBytes)
+			}
+
+			// Update stats.
+			cacheHit := !req.FinishWrite
+			if !cacheHit {
+				atomic.AddInt64(&u.stats.Streamed.Bytes, item.Digest.SizeBytes)
```

Comment:
This way we have no way of measuring the impact of compression on the network. 

Particularly, we should also be able to count how often we're retrying (it'd also be nice to be able to infer how many bytes we're retrying, but I think realistically it'd either be too few to matter, or so much that we'd have to fix other network/backend load issues).

Reply:
I agree. In my original version I was actually adding `WriteOffset` and unconditionally, but then realized that this is inconsistent with stats.Batched. Also it is weird to see streamed.Bytes higher than cacheMisses - it may happen if all streams resulted in cache hits.

WDYT about a new stats field dedicated specifically to the network? It would probably combine batched and streamed writes?

Reply:
a TODO was added.